lib/ansible/plugins/connection/ssh.py

```diff
@@ -309,8 +309,49 @@ def _examine_output(self, source, state, b_chunk, sudoable):
 
         return b''.join(output), remainder
 
+
     def _run(self, cmd, in_data, sudoable=True):
         '''
+        Will retry execution if ssh were unable to connect or timeout
+        '''
+        remaining_tries = int(C.ANSIBLE_SSH_RETRIES) + 1
+        cmd_summary = ' '.join(cmd)
+        for attempt in range(remaining_tries):
+            try:
+                return_tuple = self._run_once(cmd=cmd, in_data=in_data, sudoable=sudoable)
+                # 0 = success
+                # 1-254 = remote command return code
+                # 255 = failure from the ssh command itself
+                if return_tuple[0] != 255:
+                    break
+                else:
```

Comment:
_run_once IS raising AnsibleConnectionFailure is return code is 255, but only if in_data were present:
```
        if p.returncode == 255 and in_data:
            raise AnsibleConnectionFailure('SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh')
```
I guess that was kind of precaution. I intentionally didn't change any internal logic